pageserver/src/pgdatadir_mapping.rs

```diff
@@ -0,0 +1,1351 @@
+                bail!("unexpected pending WAL record");
+            }
+        } else {
+            let last_lsn = self.tline.get_last_record_lsn();
+            self.tline.tline.get(key, last_lsn)
+        }
+    }
+
+    fn put(&mut self, key: Key, val: Value) {
+        self.pending_updates.insert(key, val);
+    }
+
+    fn delete(&mut self, key_range: Range<Key>) {
+        trace!("DELETE {}-{}", key_range.start, key_range.end);
+        self.pending_deletions.push(key_range);
+    }
+}
+
+//--- Metadata structs stored in key-value pairs in the repository.
+
+#[derive(Debug, Serialize, Deserialize)]
+struct DbDirectory {
+    // (spcnode, dbnode) -> (do relmapper and PG_VERSION files exist)
+    dbdirs: HashMap<(Oid, Oid), bool>,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct TwoPhaseDirectory {
+    xids: HashSet<TransactionId>,
+}
+
+#[derive(Debug, Serialize, Deserialize, Default)]
+struct RelDirectory {
+    // Set of relations that exist. (relfilenode, forknum)
+    //
+    // TODO: Store it as a btree or radix tree or something else that spans multiple
+    // key-value pairs, if you have a lot of relations
+    rels: HashSet<(Oid, u8)>,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct RelSizeEntry {
+    nblocks: u32,
+}
+
+#[derive(Debug, Serialize, Deserialize, Default)]
+struct SlruSegmentDirectory {
+    // Set of SLRU segments that exist.
+    segments: HashSet<u32>,
+}
+
+static ZERO_PAGE: Bytes = Bytes::from_static(&[0u8; pg_constants::BLCKSZ as usize]);
+
+// Layout of the Key address space
+//
+// The Key struct, used to address the underlying key-value store, consists of
+// 18 bytes, split into six fields. See 'Key' in repository.rs. We need to map
+// all the data and metadata keys into those 18 bytes.
+//
+// Principles for the mapping:
+//
+// - Things that are often accessed or modified together, should be close to
+//   each other in the key space. For example, if a relation is extended by one
+//   block, we create a new key-value pair for the block data, and update the
+//   relation size entry. Because of that, the RelSize key comes after all the
+//   RelBlocks of a relation: the RelSize and the last RelBlock are always next
+//   to each other.
+//
+// The key space is divided into four major sections, identified by the first
+// byte, and the form a hierarchy:
+//
+// 00 Relation data and metadata
+//
+//   DbDir    () -> (dbnode, spcnode)
+//   Filenodemap
+//   RelDir   -> relnode forknum
+//       RelBlocks
+//       RelSize
+//
+// 01 SLRUs
+//
+//   SlruDir  kind
+//   SlruSegBlocks segno
+//   SlruSegSize
+//
+// 02 pg_twophase
+//
+// 03 misc
+//    controlfile
+//    checkpoint
+//
+// Below is a full list of the keyspace allocation:
+//
+// DbDir:
+// 00 00000000 00000000 00000000 00   00000000
+//
+// Filenodemap:
+// 00 SPCNODE  DBNODE   00000000 00   00000000
+//
+// RelDir:
+// 00 SPCNODE  DBNODE   00000000 00   00000001
```

Comment:
Here we rely on the Postgres-specific knowledge that relnode is never 0. 
Let's mention it and other assumptions in the comment.